crates/runtime/src/externref.rs

```diff
@@ -614,17 +614,39 @@ impl VMExternRefActivationsTable {
         }
     }
 
-    fn insert_precise_stack_root(&self, root: NonNull<VMExternData>) {
-        let mut precise_stack_roots = self.precise_stack_roots.borrow_mut();
+    fn insert_precise_stack_root(
+        precise_stack_roots: &mut HashSet<VMExternRefWithTraits>,
+        root: NonNull<VMExternData>,
+    ) {
         let root = unsafe { VMExternRef::clone_from_raw(root.as_ptr() as *mut _) };
         precise_stack_roots.insert(VMExternRefWithTraits(root));
     }
 
     /// Sweep the bump allocation table after we've discovered our precise stack
     /// roots.
-    fn sweep(&self) {
+    fn sweep(&self, precise_stack_roots: &mut HashSet<VMExternRefWithTraits>) {
+        // Swap out the over-approximated set so we can distinguish between the
+        // over-approximation before we started sweeping, and any new elements
+        // we might insert into the table because of re-entering Wasm via an
+        // `externref`'s destructor. The new elements must be kept alive for
+        // memory safety, but we keep this set around because we likely want to
+        // reuse its allocation/capacity for the new `precise_stack_roots` in
+        // the next GC cycle.
+        let mut old_over_approximated = mem::replace(
+            &mut *self.over_approximated_stack_roots.borrow_mut(),
+            Default::default(),
+        );
+
         // Sweep our bump chunk.
+        //
+        // Just in case an `externref` destructor calls back into Wasm, passing
+        // more `externref`s into that Wasm, which requires the `externref`s to
+        // be inserted into this `VMExternRefActivationsTable`, make sure `next
+        // == end` so that they go into the over-approximation hash set.
```

Comment:
If GC/wasm happens multiple times, will this property always hold though? This seems like it'd need to both set next == end but also some sort of flag saying "always take the slow path, never reset next/end"

Reply:
If you try to insert into the activations table, and `next == end`, then you do a gc (which will be re-entrant in this case, and exit early) and then insert into the over-approximized hash set. So, yes this relies on no one but the GC resetting next/end, but that property does hold right now.

Reply:
Oh right sorry I thought this happened somewhere in `try_insert` or similar, but it actually happens in this function! In that case the recursive guard on `gc` should be good enough 👍